core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/OptimizeJoinsBelowEmptyAggregationRule.java

```diff
@@ -0,0 +1,156 @@
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isOptimizeOutputDuplicatesInsensitiveJoins(session);
+    }
+
+    @Override
+    public Result apply(AggregationNode aggregation, Captures captures, Context context)
+    {
+        return aggregation.getSource().accept(new Rewriter(metadata, context.getLookup()), null)
+                .map(rewrittenSource -> Result.ofPlanNode(aggregation.replaceChildren(ImmutableList.of(rewrittenSource))))
+                .orElse(Result.empty());
+    }
+
+    private static class Rewriter
+            extends PlanVisitor<Optional<PlanNode>, Void>
+    {
+        private final Metadata metadata;
+        private final Lookup lookup;
+
+        private Rewriter(Metadata metadata, Lookup lookup)
+        {
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.lookup = requireNonNull(lookup, "lookup is null");
+        }
+
+        @Override
+        protected Optional<PlanNode> visitPlan(PlanNode node, Void context)
+        {
+            return Optional.empty();
+        }
+
+        @Override
+        public Optional<PlanNode> visitFilter(FilterNode node, Void context)
+        {
+            if (!isDeterministic(node.getPredicate(), metadata)) {
+                // non-deterministic expressions could filter duplicate rows probabilistically
+                return Optional.empty();
+            }
+
+            return node.getSource().accept(this, null)
+                    .map(source -> node.replaceChildren(ImmutableList.of(source)));
+        }
+
+        @Override
+        public Optional<PlanNode> visitProject(ProjectNode node, Void context)
+        {
+            boolean isDeterministic = node.getAssignments().getExpressions().stream()
+                    .allMatch(expression -> isDeterministic(expression, metadata));
+            if (!isDeterministic) {
+                // non-deterministic projections could be used in downstream filters which could
+                // filter duplicate rows probabilistically
+                return Optional.empty();
+            }
+
+            return node.getSource().accept(this, null)
+                    .map(source -> node.replaceChildren(ImmutableList.of(source)));
+        }
+
+        @Override
+        public Optional<PlanNode> visitJoin(JoinNode node, Void context)
+        {
+            if (!isDeterministic(node.getFilter().orElse(TRUE_LITERAL), metadata)) {
+                // LookupJoinOperator will evaluate non-deterministic condition on output rows until one of the
+                // rows matches. Therefore it's safe to set maySkipOutputDuplicates for joins with non-deterministic
+                // filters.
+                if (node.isMaySkipOutputDuplicates()) {
+                    return Optional.empty();
+                }
```

Comment:
I'm not sure I understand this. If the join node is already marked with `maySkipOutputDuplicates`, why would we not preserve it regardless? If a join with a non-deterministic non equi condition is incompatible with isMaySkipOutputDuplicates, we would've never got in this state (and we should validate that somewhere else -- e.g., when pushing filters into join and/or when constructing a join node)